### PR TITLE
Newly added cases to cover:do blockresize for device which has backing chain element

### PR DESCRIPTION
**Pull request by chunfuwen**
Newly added cases to cover:do blockresize for device which has backing chain element

Create external disk-only snapshot, resizing block device with backing chain
should work as expected

Signed-off-by: chunfuwen <chwen@redhat.com>

### CODE REVIEW

**Comment by chunfuwen** [comment]
Command 'virsh list' finished with 0 after 0.033422339s
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.snapshot_blockresize.positive_test.backing_chain_element

File: libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    try:
+        # Start VM
+        if vm.is_dead():
+            vm.start()
+        vm.wait_for_login().close()
+        # Cleanup snapshots if exists
+        libvirt_disk.cleanup_snapshots(vm, external_snapshot_disks)
+        external_snapshot_disks = libvirt_disk.make_external_disk_snapshots(vm, device_target, snapshot_name, snapshot_take)
+    except virt_vm.VMStartError as e:
+        if status_error:
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:59]: Currently there is only one case with  status_error = "no", so line 59 ~ 61 is redundant so far. Please remove them.

File: libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    5.Clean up test environment
+    """
+    vm_name = params.get("main_vm")
+    vm = env.get_vm(vm_name)
+
+    def check_resized_block():
+        """
+        Create external disk-only snapshots
+
+        :param params: one dict to wrap up parameters
```
Comment (dzhengfy) [inline libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:30]: No param parameter

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

**Comment by nanli1** [comment]
@chunfuwen Hi chunfu, just confirm. Could you please give which polarion case matched? Because the whole process of this  pr may be almost same as  https://github.com/autotest/tp-libvirt/pull/4047  which matched RHEL-133310. If duplicated, we may need to discuss with feature owner.

**Review by nanli1 — COMMENTED** [review]
@chunfuwen Hi chunfu, just confirm. Could you please give which polarion case matched? Because the whole process of this pr may be almost same as https://github.com/autotest/tp-libvirt/pull/4047 which matched RHEL-133310. If duplicated, we may need to discuss with feature owner.

**Comment by chunfuwen** [comment]
> @chunfuwen Hi chunfu, just confirm. Could you please give which polarion case matched? Because the whole process of this pr may be almost same as #4047 which matched RHEL-133310. If duplicated, we may need to discuss with feature owner.

@nanli1 ,it is xx-294598

**Comment by nanli1** [comment]
> > @chunfuwen Hi chunfu, just confirm. Could you please give which polarion case matched? Because the whole process of this pr may be almost same as #4047 which matched RHEL-133310. If duplicated, we may need to discuss with feature owner.
> 
> @nanli1 ,it is xx-294598

@meinaLi hi meina ,Could you please help check if this two case duplicated or not - xx-294598 and RHEL-133310

**Comment by meinaLi** [comment]
@nanli1 @chunfuwen  Yes, they are the same cases. The only difference is, VIRT-294598 is the modular case which is removed from the backingchain feature to the virtual disk feature. RHEL-13310 is the old backingchain case.

**Comment by nanli1** [comment]
> @nanli1 @chunfuwen Yes, they are the same cases. The only difference is, VIRT-294598 is the modular case which is removed from the backingchain feature to the virtual disk feature. RHEL-13310 is the old backingchain case.


> @nanli1 @chunfuwen Yes, they are the same cases. The only difference is, VIRT-294598 is the modular case which is removed from the backingchain feature to the virtual disk feature. RHEL-13310 is the old backingchain case.

OK ,Thanks meina.

File: libvirt/tests/cfg/virtual_disks/virtual_disks_snapshot_blockresize.cfg [hunk 3]
```diff
@@ -0,0 +1,14 @@
+- virtual_disks.snapshot_blockresize:
+    type = virtual_disks_snapshot_blockresize
+    take_regular_screendumps = "no"
+    start_vm = "no"
```
[the remark this answers is not in the document]
Comment (chunfuwen) [reply libvirt/tests/cfg/virtual_disks/virtual_disks_snapshot_blockresize.cfg:4]: remove them

File: libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py (hunk 1 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:59]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py (hunk 2 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:30]: updated

**Comment by chunfuwen** [comment]
new result after code is updated

Command 'virsh list' finished with 0 after 0.033642311s
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virtual_disks.snapshot_blockresize.positive_test.backing_chain_element

File: libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    # Back up xml file.
+    vmxml_backup = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    try:
+        vm.wait_for_login().close()
+        # Cleanup snapshots if exists
+        libvirt_disk.cleanup_snapshots(vm, external_snapshot_disks)
+        external_snapshot_disks = libvirt_disk.make_external_disk_snapshots(vm, device_target, snapshot_name, snapshot_take)
+    except virt_vm.VMStartError as e:
+        test.fail("VM failed to start."
+                  "Error: %s" % str(e))
```
Comment (nanli1) [inline libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:56]: Just confirm , Is this error message correct?why it's "Failed to **start** " it should be "Failed to login or ***"

Comment (nanli1) [inline libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:56]: Just confirm , Is this error message correct?why it's`  "Failed to **start** " `it should be` "Failed to login or ***"`

File: libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    else:
+        check_resized_block()
+    finally:
+        # Clean up snapshots
+        libvirt_disk.cleanup_snapshots(vm, external_snapshot_disks)
+        if vm.is_alive():
+            vm.destroy(gracefully=False)
```
Comment (nanli1) [inline libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:63]: Do we need to destroy vm after test ? vm state should be shutoff even if you not destroy it , right?

Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:63]: this is common pattern that we need destroy vm before calling later vmxml_backup.sync(). 
You can notice this almost python files in tp-libvirt

**Comment by chunfuwen** [comment]
@dzhengfy  comments are replied

File: libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py (hunk 5 above)
Comment (nanli1) [reply libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:63]: Oh....Due to my bad understanding?.
I find many files not do this  https://github.com/autotest/tp-libvirt/blob/master/libvirt/tests/src/cpu/vm_features.py 
https://github.com/autotest/tp-libvirt/blob/master/libvirt/tests/src/cpu/max_vcpus.py  https://github.com/autotest/tp-libvirt/blob/master/libvirt/tests/src/memory/memory_hotplug.py

Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_snapshot_blockresize.py:63]: @nanli1 
if you start VM in your code, it is very natural to destroy it at the end. it is common pattern whether it is VM or other resources.
In some situation, maybe some framework such as CI(downstream infrastructure) may help destroy VM if code is not handled properly itself. But it is upstream code, and tp-libvirt is standalone project itself ,and it should not depend on downstream  infrastructure.

**Review by nanli1 — APPROVED** [review]

**Review by dzhengfy — APPROVED** [review]
LGTM

**Pull request merged**